### ISSUE TITLE
$guardHandler->authenticateUserAndHandleSuccess()  not working

### ISSUE THREAD

**Issue by ThomasLandauer** [symfony/symfony#29784]
**Symfony version(s) affected**: 4.2.1

**Description**  
This is not working as expected. The user is redirected to the login form, not to the restricted area:
```php
$guardHandler->authenticateUserAndHandleSuccess($singer, $request, $authenticator, 'singer_firewall');
```
However, if I do it "manually", it works as expected (i.e. user redirected to restricted area):
```php
$token = new UsernamePasswordToken($singer, null, 'singer_firewall', $singer->getRoles());
$tokenStorage->setToken($token);
$session = $request->getSession();
$session->set('_security_singer_firewall', serialize($token));
$event = new \Symfony\Component\Security\Http\Event\InteractiveLoginEvent($request, $token);
$eventDispatcher->dispatch("security.interactive_login", $event);
return $this->redirectToRoute('singer_profile', array('singerId'=>$singer->getId()));
```
That's why I'm positive that my config is correct ;-)  
Wasn't `authenticateUserAndHandleSuccess()` meant to be a "shortcut" for this?

**Additional context**  
* I have multiple firewalls.
* My `onAuthenticationSuccess()` returns a `RedirectResponse` (basically the same as above).
* If I `die;` in `onAuthenticationSuccess()`, Symfony's debug toolbar shows me that I am authenticated. However, there's no "Firewall name" shown in the popup.

**Comment by xabbuh** [comment]
Can you create a small example application that allows to reproduce?

**Comment by ThomasLandauer** [comment]
Hurray, this is my first Symfony bug and also my first GitHub repository :-) I followed the instructions at https://symfony.com/doc/current/contributing/code/reproducer.html and here's the result: https://github.com/ThomasLandauer/symfony-issue-29784

The problem comes from this line in `security.yaml`:
```yaml
pattern: ^/(user|login)
```

Instructions:
1. Go to http://localhost:8000/register and submit the form with anything (but remember it). => You get redirected to `/login` (not to `/user/{id}` - this is the bug).
2. To double-check that security is set up correctly, enter the login info from above. => You get redirected to `/user/{id}` - fine!
3. Open `config/packages/security.yaml` and remove the line `pattern: ^/(user|login)`
4. Delete PHP's session temp save directory (`/var/lib/php/sessions` or whatever it is for you), to make sure you really start from scratch.
5. Repeat step 1 (be sure to refresh before you submit, for the CSRF token). => You do get redirected to `/user/{id}`. Bingo.

**Comment by ThomasLandauer** [comment]
OK, so the solution is to just append `|register` to the pattern?! I'm sure I tried that yesterday, but it didn't work - however, it works today ;-)

Anyway, back to the bug. If I keep the `^/(user|login)` pattern and `echo` something in `onAuthenticationSuccess()`, I see that it's being executed after registration. Who calls that??

In any case, all this should be explained in the docs - just mentioning this here as a reminder: https://symfony.com/doc/current/security/guard_authentication.html#manually-authenticating-a-user

**Comment by xabbuh** [comment]
By the way, I did some more debugging and with a fresh session I get redirected to `/user/<id>`, but then the request is not handled by your `LoginFormAuthenticator` due to its implementation of the `supports()` method:

```php
return 'app_login' === $request->attributes->get('_route')
            && $request->isMethod('POST');
```

**Comment by ThomasLandauer** [comment]
OK, so `authenticateUserAndHandleSuccess()` calls `onAuthenticationSuccess()`. But why do I need to append `|register` to the pattern? I thought that it's the very purpose of this direct call to avoid the routing through the entire firewall/authenticator?

And then, after the redirect which is sent by `onAuthenticationSuccess()`, the *subsequent* request is routed through the authenticator just normally, so `supports()` must also catch `GET` requests?

**Comment by weaverryan** [comment]
From @xabbuh 

> When the `RegistrationController::register()` controller is processed, the current request path is `/register`. This path does not match the `^/(user|login)` pattern. This means that this request is not processed in the context of the `main` firewall.

That's the core of the problem. Basically, when you're under `/register`, there are NO firewalls active. And each firewall behaves like an independent security system. So basically, Guard attempts to authenticate you on whatever firewall is active when you are currently on `/register` - it tries to authenticate you on the current firewall (there is actually none). So now, if you go back to, for example, `/login`, you are under the `main` firewall - but you were not authenticated under that firewall on the previous request. Or worse, if you try to go to `/admin`, that is also under NO firewall at all, so it also won't be authenticated.

What you really need is just to remove the `pattern` entirely from your `main` firewall so that it matches everything. It's not that you want to avoid "running through the entire" firewall - you ALWAYS want your firewall active. The idea is that you want to directly authenticate a person without "running through the entire authenticator".

Hope that makes sense :). It is potentially odd that you didn't get an exception when trying to authenticate via Guard when *no* firewall is active (cause that makes no sense) - so that could indeed be a bug.

Cheers!

**Comment by xabbuh** [comment]
@ThomasLandauer Are things more clear now or do we still need to do anything?

**Comment by xabbuh** [comment]
I am going to close here for now due to the lack of feedback. Please let us know when you have more information and we can consider to reopen.

**Comment by jlefils** [comment]
I have the same issue as you, if I use $guardHandler->authenticateUserAndHandleSuccess(), my user is only authenticated to the actual page.
Tokens are not the same (UsernamePasswordToken vs GuardToken), maybe that's the reason.
Maybe Symfony force users to authenticate from the form only, but the documentation and examples seems to say the contrary... :)
Does someone knows if we can or can't authenticate a user after registration or do we have to redirect to login form in Symfony 5?
Thks guys !

**Comment by psylosss** [comment]
Same issue. Update symfony to 5.0.8, it works there.

**Comment by wouterj** [comment]
See #36208 the bug is fixed in 5.0.8 and 4.4.8
